### PR TITLE
`deer` create built-in error types

### PR DESCRIPTION
**Pull request by indietyp**
## 🌟 What is the purpose of this PR?

Create all built-in error types, which will either be used by `deer` or are expected to be common in external code. These types take inspiration from [serde](https://docs.rs/serde/latest/serde/de/trait.Error.html), because `deer` does not necessarily intend to support parsing (which is supposed to be handed over to other crates, like `serde`). Some types have been omitted (like duplicate fields).

## 🔗 Related links

- [Asana task](https://app.asana.com/0/0/1203363157432057) _(internal)_

## 📹 Demo

image

### CODE REVIEW

File: packages/libs/deer/src/error/macros.rs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+macro_rules! id {
+    [$($segment:literal),*] => {
+        $crate::error::Id::new(&[$($segment),*])
+    };
+}
+
+macro_rules! impl_error {
+    ($name:ident) => {
+        #[cfg(nightly)]
+        impl core::error::Error for $name {}
```
Comment (TimDiekmann) [inline packages/libs/deer/src/error/macros.rs:11]: Can't wait that `core::error::Error` is stabilized 🙃

File: packages/libs/deer/src/error/mod.rs [hunk 2]
```diff
@@ -83,6 +104,78 @@ impl Id {
     }
 }
 
+// TODO: most likely (in 0.2) we want to actually have a proper schema
+pub struct Schema {
+    ty: String,
+    other: BTreeMap<String, Box<dyn erased_serde::Serialize + Send + Sync>>,
```
Comment (TimDiekmann) [inline packages/libs/deer/src/error/mod.rs:110]: Is there a specific reason to use `BTreeMap` over `HashMap`?

File: packages/libs/deer/src/error/mod.rs [hunk 3]
```diff
@@ -83,6 +104,78 @@ impl Id {
     }
 }
 
+// TODO: most likely (in 0.2) we want to actually have a proper schema
+pub struct Schema {
+    ty: String,
```
Comment (TimDiekmann) [inline packages/libs/deer/src/error/mod.rs:109]: I wonder, if we want `String` or `Cow<str>` to support non-owned values, but probably it's not worth the hassle as we have a big struct next to it anyway.

File: packages/libs/deer/src/error/mod.rs [hunk 4]
```diff
@@ -83,6 +104,78 @@ impl Id {
...
+
+    pub fn set(
+        &mut self,
+        key: impl Into<String>,
+        value: impl erased_serde::Serialize + Send + Sync + 'static,
+    ) -> &mut Self {
+        self.other.insert(key.into(), Box::new(value));
+
+        self
+    }
```
Comment (TimDiekmann) [inline packages/libs/deer/src/error/mod.rs:145]: That's an interesting pattern, I like it!

File: packages/libs/deer/src/error/mod.rs [hunk 5]
```diff
@@ -83,6 +104,78 @@ impl Id {
...
+
+        for (key, value) in &self.other {
+            map.serialize_entry(key, value)?;
+        }
+
+        map.end()
+    }
+}
+
+pub(crate) fn fold_field<'a>(it: impl Iterator<Item = &'a str>) -> String {
```
Comment (TimDiekmann) [inline packages/libs/deer/src/error/mod.rs:165]: Did you consider Instead of folding using the `Formatter` directly in the implementation by passing it from the message function?

File: packages/libs/deer/src/error/mod.rs [hunk 6]
```diff
@@ -83,6 +104,78 @@ impl Id {
...
+    it.enumerate().fold(String::new(), |mut acc, (idx, field)| {
+        if idx > 0 {
+            acc.push_str(", ");
+        }
+
+        acc.push('"');
+        acc.push_str(field);
+        acc.push('"');
```
Comment (TimDiekmann) [inline packages/libs/deer/src/error/mod.rs:173]: This would then just become `write!(fmt, "\"{field}\"")?;`.

File: packages/libs/deer/src/error/tuple.rs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        $name!(T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11);
+        $name!(T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12);
+        $name!(T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13);
+        $name!(T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14);
+        $name!(T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15);
+        $name!(T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16);
+    };
+}
+
+macro_rules! properties {
```
Comment (TimDiekmann) [inline packages/libs/deer/src/error/tuple.rs:28]: Could you please add documentation on what this is doing? It's a little bit hard to parse and macro rules expansion typically doesn't have good IDE support, which would help parse it.

File: packages/libs/deer/src/error/type.rs [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+impl ErrorProperty for ReceivedType {
+    type Value<'a> = Option<&'a Self>;
+
+    fn key() -> &'static str {
+        "received"
+    }
+
+    fn value<'a>(mut stack: impl Iterator<Item = &'a Self>) -> Self::Value<'a> {
+        stack.next()
+    }
```
Comment (TimDiekmann) [inline packages/libs/deer/src/error/type.rs:59]: Isn't this just the same as `Iterator::next`?

**Review by TimDiekmann — CHANGES_REQUESTED** [review]

File: packages/libs/deer/src/error/mod.rs (hunk 2 above)
Comment (indietyp) [reply packages/libs/deer/src/error/mod.rs:110]: Yes, `HashMap` is only available on `no-std` over the hashbrown crate, while `BTreeMap` is available on `alloc`.

File: packages/libs/deer/src/error/mod.rs [hunk 9]
```diff
@@ -83,6 +104,78 @@ impl Id {
...
+
+        map.serialize_entry("type", &self.ty)?;
+
+        for (key, value) in &self.other {
+            map.serialize_entry(key, value)?;
+        }
+
+        map.end()
+    }
+}
```
[the remark this answers is not in the document]
Comment (indietyp) [reply packages/libs/deer/src/error/mod.rs]: fixed in [874d5fb](https://github.com/hashintel/hash/pull/1375/commits/874d5fb1f88adab64aa9c01c907bbcb0ecb0c115)

File: packages/libs/deer/src/error/type.rs (hunk 8 above)
Comment (indietyp) [reply packages/libs/deer/src/error/type.rs:59]: Yes, but that isn't always the case. Some implementations might want to take into account other values. For example, `Location` renders a list of all items, but one could also imagine the possibility of future types of filtering items, etc.

File: packages/libs/deer/src/error/mod.rs (hunk 5 above)
Comment (indietyp) [reply packages/libs/deer/src/error/mod.rs]: nice catch! fixed in [c42234e](https://github.com/hashintel/hash/pull/1375/commits/c42234ed02fb7f2d0523c22e238926ed4d1f264b)

File: packages/libs/deer/src/error/mod.rs (hunk 6 above)
Comment (indietyp) [reply packages/libs/deer/src/error/mod.rs]: fixed in [c42234e](https://github.com/hashintel/hash/pull/1375/commits/c42234ed02fb7f2d0523c22e238926ed4d1f264b)

File: packages/libs/deer/src/error/tuple.rs (hunk 7 above)
Comment (indietyp) [reply packages/libs/deer/src/error/tuple.rs:62]: fixed in [a072aec](https://github.com/hashintel/hash/pull/1375/commits/a072aec433437f02138c7857a21568bc72395d96)

File: packages/libs/deer/src/error/mod.rs (hunk 3 above)
Comment (indietyp) [reply packages/libs/deer/src/error/mod.rs:111]: I also thought of that, but because `other` is already a full-blown `BTreeMap`, and you likely define more than just the type, it wasn't worth the effort (for now).

**Review by TimDiekmann — APPROVED** [review]
